src/TokenRefresher.ts

```diff
@@ -0,0 +1,185 @@
+
+            return this.refreshTokens().catch(error => this.adapter.log.error(`Cannot refresh tokens: ${error}`));
+        });
+    }
+
+    /**
+     * Destroys the TokenRefresher instance, clearing any timeouts and stopping state subscriptions.
+     */
+    destroy(): void {
+        if (this.refreshTokenTimeout) {
+            this.adapter.clearTimeout(this.refreshTokenTimeout);
+            this.refreshTokenTimeout = undefined;
+        }
+    }
+
+    /**
+     * This method is called when the state changes for the token.
+     *
+     * @param id ID of the state that changed
+     * @param state Value
+     */
+    onStateChange(id: string, state: ioBroker.State | null | undefined): void {
+        if (state?.ack && id.endsWith(`.${this.stateName}`)) {
+            if (JSON.stringify(this.accessToken) !== state.val) {
+                try {
+                    this.accessToken = JSON.parse(state.val as string);
+                    this.refreshTokens().catch(error => this.adapter.log.error(`Cannot refresh tokens: ${error}`));
+                } catch (error) {
+                    this.adapter.log.error(`Cannot parse tokens: ${error as Error}`);
+                    this.accessToken = undefined;
+                }
+            }
+        }
+    }
+
+    /** Returns the access token if it is valid and not expired.*/
+    async getAccessToken(): Promise<string | undefined> {
+        await this.readyPromise;
+        if (!this.accessToken?.access_token) {
+            this.adapter.log.error(`No tokens for ${this.name} found`);
+            return undefined;
+        }
+        if (
+            !this.accessToken.access_token_expires_on ||
+            new Date(this.accessToken.access_token_expires_on).getTime() < Date.now()
+        ) {
+            this.adapter.log.error('Access token is expired. Please make a authorization again');
+            return undefined;
+        }
+        return this.accessToken.access_token;
+    }
+
+    private async refreshTokens(): Promise<void> {
+        if (this.refreshTokenTimeout) {
+            this.adapter.clearTimeout(this.refreshTokenTimeout);
+            this.refreshTokenTimeout = undefined;
+        }
+
+        if (!this.accessToken?.refresh_token) {
+            this.adapter.log.error(`No tokens for ${this.name} found`);
+            return;
+        }
+
+        if (
+            !this.accessToken.access_token_expires_on ||
+            new Date(this.accessToken.access_token_expires_on).getTime() < Date.now()
+        ) {
+            this.adapter.log.debug('Access token is expired. Retrying to refresh tokens...');
+        }
+
+        let expiresIn = new Date(this.accessToken.access_token_expires_on).getTime() - Date.now() - 180_000;
+
+        // If expiration is in less than 3 minutes, refresh the token
+        if (expiresIn <= 0) {
+            // Refresh token
+            try {
+                const response = await axios.post(this.url, this.accessToken);
```

Comment:
whats the default communication timeout to make sure it does not hang? Maybe define explicitely